File: projects/6/FullAssembler/internal/parser/parser.go
package parser

import (
	"fmt"
	"os"
	"regexp"
	"strings"
)

const (
	A_INSTRUCTION = "A_INSTRUCTION"
	C_INSTRUCTION = "C_INSTRUCTION"
	L_INSTRUCTION = "L_INSTRUCTION"
)

type IParser interface {
	CurrentLine() string
	Reset()
	HasMoreLines() bool
	Advance() error
	InstructionType() string
	Symbol() string
	Dest() string
	Comp() string
	Jump() string
}

type parser struct {
	lines            []string
	currentLineIndex int
	currentLine      string
	instructionType  string
	symbol           string
	dest             string
	comp             string
	jump             string
}

var (
	aCommandRegex = regexp.MustCompile(`^@([A-Za-z_.$:][A-Za-z0-9_.$:]*|\d+)$`)
	lCommandRegex = regexp.MustCompile(`^\(([A-Za-z_.$:][A-Za-z0-9_.$:]*)\)$`)
	cCommandRegex = regexp.MustCompile(
		`^(?:(?P<dest>[AMD]{1,3})=)?(?P<comp>0|1|-1|D|A|M|!D|!A|!M|-D|-A|-M|D\+1|A\+1|M\+1|D-1|A-1|M-1|D\+A|D\+M|D-A|D-M|A-D|M-D|D&A|D&M|D\|A|D\|M)(?:;(?P<jump>JGT|JEQ|JGE|JLT|JNE|JLE|JMP))?$`,
	)
)

func NewParser(assemblyFilePath string) (IParser, error) {
	content, err := os.ReadFile(assemblyFilePath)
	if err != nil {
		return nil, fmt.Errorf("failed to read file '%s': %w", assemblyFilePath, err)
	}

	if len(content) == 0 {
		return &parser{lines: []string{}}, nil
	}

	lines := strings.Split(strings.ReplaceAll(string(content), "\r\n", "\n"), "\n")

	return &parser{
		lines:            lines,
		currentLineIndex: 0,
	}, nil
}

func (p *parser) CurrentLine() string {
	return p.currentLine
}

func (p *parser) Reset() {
	p.currentLineIndex = 0
}

func (p *parser) HasMoreLines() bool {
	return p.currentLineIndex < len(p.lines)
}

func (p *parser) Advance() error {
	p.resetInstructionParts()
	for p.currentLineIndex < len(p.lines) {
		line := p.lines[p.currentLineIndex]
		p.currentLineIndex++

		line = strings.TrimSpace(line)

		if commentIndex := strings.Index(line, "//"); commentIndex >= 0 {
			line = strings.TrimSpace(line[:commentIndex])
		}

		if len(line) == 0 || strings.HasPrefix(line, "//") {
			continue
		}

		p.currentLine = line
		return p.parseInstruction(line)
	}
	return nil
}

func (p *parser) resetInstructionParts() {
	p.currentLine = ""
	p.instructionType = ""
	p.symbol = ""
	p.dest = ""
	p.comp = ""
	p.jump = ""
}

func (p *parser) parseInstruction(line string) error {
	if match := aCommandRegex.FindStringSubmatch(line); match != nil {
		p.instructionType = A_INSTRUCTION
		p.symbol = match[1]
		return nil
	}

	if match := lCommandRegex.FindStringSubmatch(line); match != nil {
		p.instructionType = L_INSTRUCTION
		p.symbol = match[1]
		return nil
	}

	if match := cCommandRegex.FindStringSubmatch(line); match != nil {
		p.instructionType = C_INSTRUCTION

		groupNames := cCommandRegex.SubexpNames()
		for i, name := range groupNames {
			if i == 0 || name == "" {
				continue
			}
			switch name {
			case "dest":
				p.dest = match[i]
			case "comp":
				p.comp = match[i]
			case "jump":
				p.jump = match[i]
			}
		}
		return nil
	}

	return fmt.Errorf("invalid instruction format on line %d: '%s'", p.currentLineIndex, line)
}

func (p *parser) InstructionType() string { return p.instructionType }

func (p *parser) Symbol() string { return p.symbol }

func (p *parser) Dest() string { return p.dest }

func (p *parser) Comp() string { return p.comp }

func (p *parser) Jump() string { return p.jump }

File: projects/6/BasicAssembler/BasicAssembler/Parser.cs
using System.Text.RegularExpressions;

namespace BasicAssembler;

public interface IParser
{
    public const string A_INSTRUCTION = "A";
    public const string C_INSTRUCTION = "C";
    public const string L_INSTRUCTION = "L";

    string CurrentLine();

    bool HasMoreLines();

    void Advance();

    string InstructionType();

    string Symbol();

    string Dest();

    string Comp();

    string Jump();
}

public class Parser : IParser
{
    private string[] _lines;
    private int _currentLineIndex;
    private string _currentLine;


    private string _instructionType;
    private string _symbol;
    private string _dest;
    private string _comp;
    private string _jump;

    private static readonly Regex _aCommandRegex =
        new(pattern: @"^@([A-Za-z_.$:][A-Za-z0-9_.$:]*|\d+)$", options: RegexOptions.Compiled);

    private static readonly Regex _lCommandRegex =
        new(pattern: @"^\(([A-Za-z_.$:][A-Za-z0-9_.$:]*)\)$", options: RegexOptions.Compiled);

    // C 命令: [dest=]comp[;jump]
    private static readonly Regex _cCommandRegex = new(
        pattern: @"
            ^
            (?:(?<dest>[AMD]{1,3})=)?
            (?<comp>
                0|1|-1
                |D|A|M
                |!D|!A|!M
                |-D|-A|-M
                |D\+1|A\+1|M\+1
                |D-1|A-1|M-1
                |D\+A|D\+M
                |D-A|D-M
                |A-D|M-D
                |D&A|D&M
                |D\|A|D\|M
            )
            (?:;(?<jump>JGT|JEQ|JGE|JLT|JNE|JLE|JMP))?
            $
        ",
        options: RegexOptions.Compiled | RegexOptions.IgnorePatternWhitespace
    );


    protected Parser(string[] assemblyLines)
    {
        _lines = assemblyLines;
        _currentLineIndex = 0;
        _currentLine = string.Empty;
        _instructionType = string.Empty;
        _symbol = string.Empty;
        _dest = string.Empty;
        _comp = string.Empty;
        _jump = string.Empty;
    }

    public static IParser CreateParser(string assemblyFilePath)
    {
        string path = assemblyFilePath;
        if (!Path.IsPathRooted(assemblyFilePath))
        {
            path = Path.Combine(Directory.GetCurrentDirectory(), assemblyFilePath);
        }
        var lines = File.ReadAllLines(path);
        return new Parser(lines);
    }

    public string CurrentLine() => _currentLine;

    public bool HasMoreLines()
    {
        return _currentLineIndex < _lines.Length;
    }

    public void Advance()
    {
        _currentLine = string.Empty;
        _instructionType = string.Empty;
        _symbol = string.Empty;
        _dest = string.Empty;
        _comp = string.Empty;
        _jump = string.Empty;

        while (HasMoreLines())
        {
            var line = _lines[_currentLineIndex++].Trim();

            // コメント行や空行をまとめてスキップ
            if (line.Length == 0 || line.StartsWith("//"))
                continue;

            int commentIndex = line.IndexOf("//");
            if (commentIndex >= 0)
                line = line[..commentIndex].Trim();

            if (line.Length == 0)
                continue;

            _currentLine = line;
            ParseInstruction(line);
            break;
        }
    }

    public string InstructionType() => _instructionType;

    public string Symbol() => _symbol;

    public string Dest() => _dest;

    public string Comp() => _comp;

    public string Jump() => _jump;

    private void ParseInstruction(string line)
    {
        var aCommandMatch = _aCommandRegex.Match(line);
        if (aCommandMatch.Success)
        {
            _instructionType = IParser.A_INSTRUCTION;
            _symbol = aCommandMatch.Groups[1].Value;
            return;
        }

        var lCommandMatch = _lCommandRegex.Match(line);
        if (lCommandMatch.Success)
        {
            _instructionType = IParser.L_INSTRUCTION;
            _symbol = lCommandMatch.Groups[1].Value;
            return;
        }

        var cCommandMatch = _cCommandRegex.Match(line);
        if (cCommandMatch.Success)
        {
            _instructionType = IParser.C_INSTRUCTION;
            _dest = cCommandMatch.Groups["dest"].Value;
            _comp = cCommandMatch.Groups["comp"].Value;
            _jump = cCommandMatch.Groups["jump"].Value;
            return;
        }

        throw new InvalidOperationException($"Invalid instruction format: {line}");
    }
}